fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java

```diff
@@ -0,0 +1,190 @@
+     * @param nextCursorGenerator the method which can generate the underlying cursor given a record context and a
+     * continuation
+     */
+    public AutoContinuingCursor(@Nonnull FDBDatabaseRunner runner,
+                                @Nonnull BiFunction<FDBRecordContext, byte[], RecordCursor<T>> nextCursorGenerator) {
+        this.runner = runner;
+        this.nextCursorGenerator = nextCursorGenerator;
+    }
+
+    @Nonnull
+    @Override
+    public CompletableFuture<RecordCursorResult<T>> onNext() {
+        if (currentCursor == null) {
+            openContextAndGenerateCursor(null);
+        }
+        return AsyncUtil.whileTrue(() ->
+                        currentCursor.onNext().thenApply(result -> {
+                            if (result.hasStoppedBeforeEnd()) {
+                                openContextAndGenerateCursor(result.getContinuation().toBytes());
+                                return true;
+                            } else {
+                                lastResult = result;
+                                return false;
+                            }
+                        }), getExecutor())
+                .thenApply(ignore -> lastResult);
+    }
+
+    private void openContextAndGenerateCursor(@Nullable byte[] continuation) {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Open context and generate a cursor");
+        }
+        if (currentContext != null) {
+            currentContext.close();
+        }
+        currentContext = runner.openContext();
+        currentCursor = nextCursorGenerator.apply(currentContext, continuation);
+    }
+
+    @Nonnull
+    @Override
+    public CompletableFuture<Boolean> onHasNext() {
+        if (nextFuture == null) {
+            mayGetContinuation = false;
+            nextFuture = onNext().thenApply(RecordCursorResult::hasNext);
+        }
+        return nextFuture;
+    }
+
+    @Nullable
+    @Override
+    public T next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        nextFuture = null;
+        mayGetContinuation = true;
+        return lastResult.get();
+    }
+
+    @Nullable
+    @Override
+    public byte[] getContinuation() {
+        IllegalContinuationAccessChecker.check(mayGetContinuation);
+        return lastResult.getContinuation().toBytes();
+    }
+
+    @Override
+    public NoNextReason getNoNextReason() {
+        return lastResult.getNoNextReason();
+    }
+
+    @Override
+    public void close() {
+        if (nextFuture != null) {
+            nextFuture.cancel(true);
+        }
+        if (currentContext != null) {
+            currentContext.close();
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return runner.getExecutor();
+    }
+
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
```

Comment:
Because `AutoContinuingCursor` has a child cursor, it should probably pass the visitor down to its (current) child. I think this method should look something like:
```java
if (visitor.visitEnter(this)) {
    currentCursor.accept(visitor);
}
return visitor.visitLeave(this);
```

Reply:
Done.

Reply:
Ack, `currentCursor` is nullable, right? I guess we need to check for that, too :(

Reply:
Updated in the same commit.